### DISCUSSION TITLE
documentation missing for sys headers

### DISCUSSION THREAD

**Discussion by extrawurst** [libgit2/libgit2#6186] [Q&A]
on my chase down the rabbit hole of how to potentially support git-lfs in gitui (-> git2-rs -> libgit2) I ended up wondering why `git_filter_register` doc-comments do not appear in the [documentation](https://libgit2.org/libgit2/#HEAD/search/git_filter_register)? or why none of the things inside `include/git2/sys/*` seem to appear?

**Comment by ethomson** [comment]
Sigh.  I'm honestly not sure. The docs have always been handled by @carlosmn or @tiennou, to my eternal thanks.

Are you interested in building your own lfs or exec'ing git-lfs itself?

**Reply by extrawurst** [reply]
A PoC is first running `git-lfs` using the the filters. I am not using git-lfs myself, reverse engineering this is already going to be fun (I did not find docs what calls I need to make and when ^^). 
I certainly wished there was a c-lib for lfs but creating one is not my priority right now :)

**Reply by ethomson** [reply]
I agree, it's disappointing the git-lfs is in go.

I know that at least one intrepid person has built their own lfs implementation but to my knowledge it's not public.

Here's a filter driver that will exec whatever is specified: https://github.com/libgit2/libgit2/pull/5507/commits/c392ad6df32ea6c0e460ce0622adce9df51598cd but it's not in a state that you could just copy paste it.  I hope to offer it as part of libgit2 soonish.

**Reply by lrm29** [reply]
Some implementation details in case you're not aware:
- For basic (no locking) git lfs support you just need to run the filters and also the pre-push hook.
- On Windows particularly, the filter process (https://git-scm.com/docs/gitattributes#_filter) turns out to be almost essential for performance, as running a filter per file is slow (say in the order of 1 second). I haven't got around to looking into that yet but it will probably come up for you at some point.

**Comment by carlosmn** [comment]
Usually missing functions is because clang/llvm didn't like the header as it's missing some include that works when you include it after `git2.h` but not when it's on its own.

**Comment by tiennou** [comment]
I have no idea why it's not showing up in the main documentation. My own run of current docurium `master` (aab073d) with my custom-written Rakefile tasks pushed this: https://tiennou.github.io/libgit2/#HEAD/group/filter/git_filter_register, so I wager the project's runner is not running master?

**Reply by ethomson** [reply]
> I wager the project's runner is not running master?

Aha, yes, you are correct @tiennou!  We're running [`gem install docurium`](https://github.com/libgit2/libgit2/blob/main/ci/docker/docurium#L5) which installs v0.6.0.  🤔

**Reply by carlosmn** [reply]
That's the latest release, I guess it'd make sense to cut a new one

**Reply by ethomson** [reply]
Alternately, we can fix the docs CI runs to pull in `main` (or some fixed hash).  I'm fine with that, I just don't know how to do it in Ruby land.  I don't think it's as easy as `gem install git@github.com:libgit2/docurium`...

**Reply by carlosmn** [reply]
Not *quite* but if you have a `Gemfile` then you can specify that you want a gem out of git and specif the branch. I've been trying to get a new version released, but llvm or clang have been trolling me by failing randomly and not getting all of the builds to succeed in one go.

**Comment by ethomson** [comment]
I'm seeing the sys headers in the latest documentation - thanks @carlosmn for publishing a new version of docurium.  Are there any types that you are still missing @extrawurst?

**Reply by extrawurst** [reply]
Hm I am not sure what I am missing but I can still not find `git_filter_register` in https://libgit2.org/libgit2/#HEAD

<img width="844" alt="Screenshot 2022-02-21 at 20 00 04" src="https://user-images.githubusercontent.com/776816/155013083-77c83913-eb99-484c-9daf-1e147c21dbb0.png">
